test/tests/functional/pbs_qstat_count.py

```diff
@@ -0,0 +1,271 @@
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from tests.functional import *
+
+
+class TestqstatStateCount(TestFunctional):
+
+    def setUp(self):
+        TestFunctional.setUp(self)
+        # set ncpus to a known value, 2 here
+        a = {'resources_available.ncpus': 2}
+        self.server.manager(MGR_CMD_SET, NODE, a,
+                            self.mom.shortname, expect=True)
+
+    def submit_waiting_job(self, timedelta):
+        """
+        Submit a job in W state using -a option.
+        The time specified in -a is current time + timedelta.
+        """
+        attribs = {ATTR_a: BatchUtils().convert_seconds_to_datetime(
+            int(time.time()) + timedelta)}
+        j = Job(TEST_USER, attribs)
+        jid = self.server.submit(j)
+        return jid
+
+    def find_state_counts(self, Running_count):
+        """
+        From the output of qstat -Bf, parses the no.of jobs in H W and Q states
+        and the value of total_jobs. Calculates the total no.of jobs based
+        on individual counts parsed. Returns these values in a dictionary.
+        """
+        counts = {}
+        # Get output of qstat
+        qstat = self.server.status(SERVER)
+        state_count = qstat[0]['state_count'].split()
+        all_state_count = 0
+        for s in state_count:
+            state = s.split(':')
+            # Check for negative value
+            self.assertGreaterEqual(
+                int(state[1]), 0, 'state count has negative values')
+            counts[state[0]] = int(state[1])
+            all_state_count = all_state_count + int(state[1])
+        counts['all_state_count'] = all_state_count
+        counts['total_jobs'] = int(qstat[0]['total_jobs'])
+        # Find queued count from output of qstat
+        counts['expected_queued_count'] = (counts['total_jobs']
+                                           - counts['Held']
+                                           - counts['Waiting']
+                                           - Running_count)
+        self.logger.info(counts)
+        return counts
+
+    def verify_count(self, Running_count):
+        """
+        Verify that total_jobs and queued_count reported by the output
+        of qstat -Bf matched the number of jobs submitted, where each
+        job uses ncpus=1.
+        """
+        counts = self.find_state_counts(Running_count)
+        self.assertEqual(counts['total_jobs'],
+                         counts['all_state_count'], 'Job count incorrect')
+        self.assertEqual(counts['expected_queued_count'], counts['Queued'],
+                         'Queued count incorrect')
+
+    def test_queued_no_restart(self):
+        """
+        The test case verifies that the reported queued_count in qstat -Bf
+        without a server restart is equal to the number of jobs
+        (each using ncpus=1) submitted.
```

Comment:
the test case verifies that qstat-reported queued_count = total_jobs - running jobs.